org.eclipse.jdt.core/compiler/org/eclipse/jdt/internal/compiler/parser/Parser.java

```diff
@@ -12772,6 +12772,20 @@ protected void markEnclosingMemberWithLocalOrFunctionalType(LocalTypeKind contex
 				case LOCAL:
 					node.bits |= ASTNode.HasLocalType;
 			}
+			// mark enclosing methods as having functional interface types, if the current node has such
+			if ((node.bits & ASTNode.HasFunctionalInterfaceTypes) != 0) {
+				List<Object> ancestors = new ArrayList<>(i);
+				ancestors.add(this.referenceContext);
+				for (int j = i - 1; j >= 0; j--) {
```

Comment:
I guess for very high nesting this might cause trouble, as would be quadratic complexity. If e.g. there is 100% nesting level and each time a lambda or a method reference is in the code.

Maybe we can stop the iteration over ancestor nodes if we see a method that is already marked with the flag... It should help reduce operations regardless of how the AST is traversed (BFS or DFS). Though as long as the flag is the same (`HasFunctionalInterfaceTypes`) I'm not sure we are allowed to do that. Since e.g. a parent method might itself already have a functional type, but its parent method still needs to be marked due to containing the node that is being visited "here".

Reply:
Seeing the AstNode.HasLocalType here (line 12773), I got an idea:

Reverting the adjustment to Parser.java and adjusting `org.eclipse.jdt.internal.core.search.indexing.SourceIndexer.purgeMethodStatements(TypeDeclaration)` like this makes the newly added tests green:

```java
	private void purgeMethodStatements(TypeDeclaration type) {
		AbstractMethodDeclaration[] methods = type.methods;
		for (int j = 0, length = methods == null ? 0 : methods.length; j < length; j++) {
			AbstractMethodDeclaration method = methods[j];
			if (method != null && (method.bits & (ASTNode.HasFunctionalInterfaceTypes | ASTNode.HasLocalType)) == 0) {
				method.statements = null;
				method.javadoc = null;
			}
		}

		TypeDeclaration[] memberTypes = type.memberTypes;
		if (memberTypes != null)
			for (int i = 0, l = memberTypes.length; i < l; i++)
				purgeMethodStatements(memberTypes[i]);
	}
```

Not sure if / how I can push this to your branch / merge request, though.

Reply:
> Not sure if / how I can push this to your branch / merge request, though.

Me neither, on idea how that works with forked repos. I believe the PR can be edited by committers/reviewers, but I don't know how that works either.

I've adjusted the change, seems like a good suggestion.